% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/align2reference.R
\name{checkAlign2ref}
\alias{checkAlign2ref}
\title{Check the alignment with the imputation reference panel}
\usage{
checkAlign2ref(
  plink,
  inputPrefix,
  referencePanel,
  bimReferenceFile,
  out2,
  out2.snp,
  out3,
  out3.snp,
  out4,
  out4.snp,
  out4.snpRetained,
  nCore = 25
)
}
\arguments{
\item{plink}{an executable program in either the current working directory
or somewhere in the command path.}

\item{inputPrefix}{the prefix of the input PLINK files.}

\item{referencePanel}{a string indicating the type of imputation 
reference panels is used: c("1000Gphase1v3_macGT1", "1000Gphase3").}

\item{bimReferenceFile}{the reference file used for the alignment, which is
a PLINK BIM alike format file.}

\item{out2}{the prefix of the output PLINK binary files after removing SNPs
whose genomic positions are not in the imputation reference,
taking SNP names into account.}

\item{out2.snp}{the output plain text file that stores the removed SNPs
whose genomic positions are not in the imputation reference,
taking SNP names into account.}

\item{out3}{the prefix of the output PLINK binary files after removing SNPs
whose genomic positions are not in the imputation reference,
ingoring SNP names.}

\item{out3.snp}{the output plain text file that stores the removed SNPs
whose genomic positions are not in the imputation reference,
ingoring SNP names.}

\item{out4}{the prefix of the output PLINK binary files after removing SNPs
whose alleles are not in the imputation reference,
taking their genomic positions into account.}

\item{out4.snp}{the output plain text file that stores the removed SNPs
whose alleles are not in the imputation reference,
taking their genomic positions into account.}

\item{out4.snpRetained}{the output plain text file that stores the
removed SNPs whose alleles are in the imputation reference,
taking their genomic positions into account.}

\item{nCore}{the number of cores used for computation.
This can be tuned along with nThread.}
}
\value{
The set of aligned PLINK files from your own study compared with
the imputation reference.
}
\description{
Perform the alignment against a reference panel by considering the following
parameters: variant name, genomic position and the allele profile.
Output files are generated sequentially depending on their previous PLINK data.
}
\details{
The output files are genrated in order. Genomic position includes
chromosomal location and base-pair position of the individual variant.
All monomorphic SNPs are retained for further processing.
}
\author{
Junfang Chen
}
