% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extend2genipe.R
\name{mergeByGenipe}
\alias{mergeByGenipe}
\title{Merge imputed files using Genipe}
\usage{
mergeByGenipe(
  inputImpute2,
  chr,
  probability,
  completionRate,
  info,
  outputPrefix
)
}
\arguments{
\item{inputImpute2}{the output from IMPUTE2.}

\item{chr}{specifiy the chromosome segment to be merged, on which the 
imputation was made.}

\item{probability}{the probability threshold for no calls. [<0.9]}

\item{completionRate}{the completion rate threshold for site exclusion. 
[<0.98]}

\item{info}{the measure of the observed statistical information associated 
with the allele frequency estimate threshold for site exclusion. (<0.00)}

\item{outputPrefix}{the prefix for the imputed output files.}
}
\value{
The merged imputed files using genipe.
}
\description{
Concatenate IMPUTE2 output files and retrieve some statistics. 
This is automatically called by the main genipe pipeline to merge IMPUTE2 
files generated for all the genomic segments. 
For details, see Genipe IMPUTE2 merger options.
}
\references{
Lemieux Perreault, L. P., et al. (2016). genipe: an automated 
genome-wide imputation pipeline with automatic reporting and statistical 
tools. Bioinformatics, 32(23), 3661-3663.
}
\author{
Junfang Chen
}
